experimental/casemapping/src/casemapping.rs

```diff
@@ -0,0 +1,149 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+#[cfg(feature = "provider_transform_internals")]
+use crate::error::Error;
+use crate::internals::*;
+#[cfg(feature = "provider_transform_internals")]
+use crate::provider::CaseMappingV1;
+use crate::provider::CaseMappingV1Marker;
+#[cfg(feature = "provider_transform_internals")]
+use icu_codepointtrie::CodePointTrieHeader;
+use icu_locid::Locale;
+use icu_provider::prelude::*;
+
+/// A struct with the ability to convert characters and strings to uppercase or lowercase,
+/// or fold them to a normalized form for case-insensitive comparison.
+///
+/// TODO: Expand documentation
+#[derive(Clone)]
+pub struct CaseMapping {
+    internals: DataPayload<CaseMappingV1Marker>,
+    locale: CaseMapLocale,
+}
+
+impl CaseMapping {
+    /// A constructor which takes a [`ResourceProvider`] and creates a [`CaseMapping`].
+    pub fn new<P>(provider: &P) -> Result<CaseMapping, DataError>
+    where
+        P: ResourceProvider<CaseMappingV1Marker> + ?Sized,
+    {
+        Self::new_with_locale(provider, &Locale::und())
+    }
+
+    /// A constructor which takes a [`ResourceProvider`] and creates a [`CaseMapping`] for the given locale.
+    pub fn new_with_locale<P>(provider: &P, locale: &Locale) -> Result<CaseMapping, DataError>
+    where
+        P: ResourceProvider<CaseMappingV1Marker> + ?Sized,
+    {
+        let internals = provider
+            .load_resource(&DataRequest::default())?
+            .take_payload()?;
+        debug_assert!(internals.get().casemap.validate().is_ok());
+        let locale = CaseMapLocale::from(locale);
+        Ok(Self { internals, locale })
+    }
+
+    /// Creates a new CaseMapping using data exported by the `icuexportdata` tool
+    /// in ICU4C. Validates that the data is consistent.
+    #[cfg(feature = "provider_transform_internals")]
+    pub fn try_from_icu(
+        trie_header: CodePointTrieHeader,
+        trie_index: &[u16],
+        trie_data: &[u16],
+        exceptions: &[u16],
+        unfold: &[u16],
+    ) -> Result<Self, Error> {
+        let internals = CaseMappingV1 {
+            casemap: CaseMappingInternals::try_from_icu(
+                trie_header,
+                trie_index,
+                trie_data,
+                exceptions,
+                unfold,
+            )?,
+        };
+        let locale = CaseMapLocale::Root;
+        Ok(Self {
+            internals: DataPayload::from_owned(internals),
+            locale,
+        })
+    }
+
+    /// Returns the lowercase mapping of the given `char`.
+    /// This function only implements simple and common mappings. Full mappings,
+    /// which can map one `char` to a string, are not included.
```

Comment:
Here and elsewhere:

Praise: Thank you for stating this caveat in the docs.

Request: Please link to the method you should use if you want the full behavior.